File: framework/languages/node/e2e/ResilienceLifecycle/Server/Provider/provider-host-factory.ts
import fs from 'node:fs';
import { Module } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { ZLinkMessageFlowLogMode } from '@zlink-systems/framework';
import {
  ZLINK_FANOUT_CLIENT,
  ZLINK_ROUTE_CLIENT,
  ZLINK_ROUTE_MESH_RUNTIME_OPTIONS,
  ZLinkModule,
  zlinkFramework
} from '@zlink-systems/nestjs';
import type {
  ZLinkFanoutClient,
  ZLinkRouteClient,
  ZLinkRouteMeshRuntimeOptions
} from '@zlink-systems/framework';
import { PacketNames, ResilienceNames } from '../../Shared/messages';
import { validateServerOptions } from './Configuration/server-options';
import type { ServerOptions } from './Configuration/server-options';
import { RESILIENCE_OPTIONS, createResilienceConfigurationModule } from '../../configuration';
import { createProviderEndpoints } from './Endpoints/provider-endpoints';
import {
  EvidenceDispatchErrorObserver,
  EvidenceRuntimeErrorSink,
  PayloadRequestHandler,
  ProfileCommandHandler,
  ProfileRequestHandler,
} from './Handlers/provider-handlers';
import { EvidenceStore } from './Infrastructure/evidence-store';
import { FaultState } from './Infrastructure/fault-state';
import { closeHttpServer, startHttpServer } from './Support/http-server';
import { createRedisLocationStore, resilienceLocationOptions } from '../../Shared/location-store';

export async function startProviderHost(): Promise<void> {
  let stopping = false;

  const ProviderModule = createProviderModule();
  const app = await NestFactory.createApplicationContext(ProviderModule, { logger: false, abortOnError: false });
  const options = app.get(RESILIENCE_OPTIONS, { strict: false }) as ServerOptions;
  const evidence = app.get(EvidenceStore, { strict: false });
  const fault = app.get(FaultState, { strict: false });
  const channel = app.get(ZLINK_ROUTE_CLIENT, { strict: false }) as ZLinkRouteClient;
  const fanout = app.get(ZLINK_FANOUT_CLIENT, { strict: false }) as ZLinkFanoutClient;
  const runtimeOptions = app.get(
    ZLINK_ROUTE_MESH_RUNTIME_OPTIONS,
    { strict: false }
  ) as ZLinkRouteMeshRuntimeOptions;
  const server = await startHttpServer(options.httpUrl, createProviderEndpoints(
    evidence,
    fault,
    channel,
    fanout,
    runtimeOptions,
    () => { stopping = true; }
  ));

  while (!stopping) {
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  await closeHttpServer(server);
  await app.close();
}

function createProviderModule(): Function {
  class ProviderModule {}
  const configuration = createResilienceConfigurationModule(validateServerOptions);

  Module({
    imports: [
      configuration,
      ZLinkModule.forRootFactory({
        imports: [configuration], inject: [RESILIENCE_OPTIONS],
        useFactory: (value: unknown) => {
          const options = value as ServerOptions;
          fs.mkdirSync(options.logDir, { recursive: true });
          const builder = zlinkFramework();
          builder
            .configureDispatch()
              .setMessageFlowObserver(EvidenceDispatchErrorObserver)
              .setRuntimeErrorSink(EvidenceRuntimeErrorSink)
              .messageFlow(ZLinkMessageFlowLogMode.KeyTransitions)
              .traceLogFile(`${options.logDir}/${options.rid}-flow.log`)
              .traceLabel(options.rid);

          if (options.redisEndpoint !== undefined && options.redisKeyPrefix !== undefined) {
            builder.addLocationStore(createRedisLocationStore({
              redisEndpoint: options.redisEndpoint,
              redisKeyPrefix: options.redisKeyPrefix
            }));
            Object.assign(builder.configureLocations(), resilienceLocationOptions());
          }
          if (options.channelEndpoint !== undefined) {
            const profile = builder.addRouteMesh('profile')
              .listen(options.channelEndpoint)
              .routingId(options.rid);
            profile.peerConnections();
            profile.channel('profile').server()
              .addRequestHandler(PacketNames.profileReq, ProfileRequestHandler)
              .addRequestHandler(PacketNames.payloadReq, PayloadRequestHandler)
              .addSendHandler(PacketNames.profileMsg, ProfileCommandHandler);
          }
          if (options.fanoutEndpoint !== undefined) {
            builder.addFanoutChannel(ResilienceNames.fanoutChannel)
              .enablePublisher(options.fanoutEndpoint);
          }
          return builder.build();
        }
      })
    ],
    providers: [
      { provide: EvidenceStore, inject: [RESILIENCE_OPTIONS], useFactory: (value: unknown) => {
        const options = value as ServerOptions; return new EvidenceStore(options.rid, options.evidenceFile);
      } },
      FaultState,
      EvidenceDispatchErrorObserver,
      EvidenceRuntimeErrorSink,
      PayloadRequestHandler,
      ProfileCommandHandler,
      ProfileRequestHandler,
    ]
  })(ProviderModule);
  return ProviderModule;
}

File: framework/languages/dotnet/e2e/ResilienceLifecycle/Server/Provider/ProviderHostFactory.cs
using Microsoft.Extensions.Configuration;

using ResilienceLifecycle.Server.Provider.Handlers;
using ResilienceLifecycle.Shared;
using Systems.Zlink;
using Zlink.Framework.AspNetCore;
using Zlink.Framework.Contracts.Dispatch;

using Zlink.Framework.Locations.Redis;
using Zlink.Framework.E2E.Diagnostics;

namespace ResilienceLifecycle.Server.Provider;

internal static class ProviderHostFactory
{
    public static WebApplication Create(string[] args)
    {
        var options = ServerOptions.Parse(args, "provider");
        Directory.CreateDirectory(options.LogDir);

        var builder = WebApplication.CreateBuilder(args);
        builder.Configuration.Sources.Clear();
        builder.Configuration.AddInMemoryCollection();
        builder.Logging.ClearProviders();
        builder.Logging.AddSimpleConsole(console =>
        {
            console.SingleLine = true;
            console.TimestampFormat = "HH:mm:ss.fff ";
        });
        builder.WebHost.UseUrls(options.HttpUrl);
        var evidence = new EvidenceStore(options.Rid, options.EvidenceFile);
        var fault = new FaultState();
        builder.Services.AddSingleton(evidence);
        builder.Services.AddSingleton(fault);
        builder.Services.AddSingleton(new E2eMessageFlowListener(
            Path.Combine(options.LogDir, $"{options.Rid}-flow.log"),
            options.Rid,
            flow =>
            {
                if (flow.Phase != "error") return;
                evidence.Add(
                    "dispatch-error"
                    + $"|surface={flow.Surface}"
                    + $"|kind={flow.MessageKind}"
                    + $"|reason={flow.Reason}"
                    + $"|action={flow.Action}"
                    + $"|packet={flow.PacketName ?? "<null>"}"
                    + $"|channel={flow.ChannelName ?? "<null>"}");
                if (fault.Mode == "observer-throws")
                    throw new InvalidOperationException("dispatch observer failure");
            }));

        builder.Services.AddZLinkFramework(framework =>
        {
            //  This E2E host is not started inside a memory-limited
            //  container. Supply a deterministic finite limit so the
            //  default Auto HWM contract does not depend on the host.
            framework.ConfigureInboundDispatch().ProcessMemoryLimitBytes =
                1UL * 1024 * 1024 * 1024;
            if (!string.IsNullOrWhiteSpace(options.RedisEndpoint))
                framework.AddLocationStore(new ZLinkRedisLocationStore(redis => { redis.ConnectionString = options.RedisEndpoint; redis.KeyPrefix = options.RedisKeyPrefix
                                  ?? throw new InvalidOperationException("Shared.RedisKeyPrefix is required."); }));
            framework.ConfigureDispatch().Diagnostics
                .SetLevel(ZLinkDiagnosticsLevel.Normal);
            var mesh = framework.AddRouteMesh(ResilienceLifecycleNames.Channel)
                .SetRoutingIdPrefix(options.Rid);
            if (string.IsNullOrWhiteSpace(options.ChannelBindHost))
                mesh.Listen(Require(options.ChannelEndpoint, "ChannelEndpoint"));
            else
                mesh.Listen(ParsePort(options.ChannelEndpoint))
                    .SetBindHost(options.ChannelBindHost);
            if (!string.IsNullOrWhiteSpace(options.ChannelAdvertiseHost))
                mesh.SetAdvertiseHost(options.ChannelAdvertiseHost);
            mesh.Channel(ResilienceLifecycleNames.Channel).Server()
                .SetWeight(options.Weight)
                .AddRequestHandler<ProfileRequestHandler, ProfileReq, ProfileRes>("ProfileReq")
                .AddSendHandler<ProfileCommandHandler, ProfileMsg>("ProfileMsg");

            if (options.ClientServerEnabled)
            {
                var server = framework.AddClientServerChannel(
                        ResilienceLifecycleNames.ClientServerChannel)
                    .Server()
                    .Listen(ParsePort(options.ClientServerEndpoint))
                    .SetBindHost(options.ClientServerBindHost ?? "127.0.0.1")
                    .SetAdvertiseHost(options.ClientServerAdvertiseHost ?? "127.0.0.1")
                    .SetWeight(options.Weight);
                server.AddRequestHandler<ProfileRequestHandler, ProfileReq, ProfileRes>("ProfileReq");
            }
        });

        var app = builder.Build();
        app.MapProviderEndpoints(options);
        return app;
    }

    private static string Require(string? value, string name)
    {
        return string.IsNullOrWhiteSpace(value)
            ? throw new InvalidOperationException($"{name} is required.")
            : value;
    }

    private static int ParsePort(string? endpoint)
    {
        if (!Uri.TryCreate(endpoint, UriKind.Absolute, out var uri)
            || uri.Port is < 1 or > 65535)
            throw new InvalidOperationException("ClientServerEndpoint must contain a valid TCP port.");
        return uri.Port;
    }
}
